napari_builtins/_qt/features_table.py

```diff
@@ -0,0 +1,544 @@
+            self.df.iat[row, col - 1] = value == Qt.Checked
+            self.dataChanged.emit(index, index, [Qt.CheckStateRole])
+            return True
+
+        if role == Qt.EditRole:
+            try:
+                if not pd.api.types.is_categorical_dtype(dtype):
+                    value = dtype.type(value)
+                self.df.iat[row, col - 1] = value
+            except ValueError:
+                return False
+            self.dataChanged.emit(index, index, [Qt.DisplayRole, Qt.EditRole])
+            return True
+
+        return False
+
+    # custom methods
+    @contextmanager
+    def changing(self):
+        self.layoutAboutToBeChanged.emit()
+        yield
+        self.layoutChanged.emit()
+
+    def replace_data(self, df):
+        with self.changing():
+            self.df = df
+
+
+class DelegateCategorical(QStyledItemDelegate):
+    """Delegate that uses comboboxes as editors for categorical data."""
+
+    def createEditor(self, parent, option, index):
+        proxy_model = index.model()
+        source_model = proxy_model.sourceModel()
+        source_index = proxy_model.mapToSource(index)
+        col = source_index.column()
+
+        dtype = source_model.df.dtypes.iat[col - 1]
+
+        if pd.api.types.is_categorical_dtype(dtype):
+            editor = QComboBox(parent)
+            categories = source_model.df.iloc[:, col - 1].cat.categories
+            editor.addItems([str(c) for c in categories])
+            # allow arrow keys selection
+            editor.setFocusPolicy(Qt.StrongFocus)
+
+            # force editor to open on first click, otherwise we need 2 clicks
+            QTimer.singleShot(0, editor.showPopup)
+            return editor
+
+        return super().createEditor(parent, option, index)
+
+    def setEditorData(self, editor, index):
+        if isinstance(editor, QComboBox):
+            value = index.model().data(index, Qt.EditRole)
+            i = editor.findText(value)
+            if i >= 0:
+                editor.setCurrentIndex(i)
+        else:
+            super().setEditorData(editor, index)
+
+    def setModelData(self, editor, model, index):
+        if isinstance(editor, QComboBox):
+            source_index = model.mapToSource(index)
+            source_model = model.sourceModel()
+            source_model.setData(
+                source_index, editor.currentText(), Qt.EditRole
+            )
+        else:
+            super().setModelData(editor, model, index)
+
+
+class BoolFriendlyProxyModel(QSortFilterProxyModel):
+    """Sort proxy model that handles booleans correctly."""
+
+    def lessThan(self, left, right):
+        left_data = self.sourceModel().data(left, Qt.EditRole)
+        right_data = self.sourceModel().data(right, Qt.EditRole)
+
+        # ensure booleans compare as expected
+        if isinstance(left_data, bool) and isinstance(right_data, bool):
+            return not left_data and right_data
```

Comment:
why is this not working? 
```suggestion
            return left_data < right_data
```

Reply:
I don't remember why I ended up on this to be honest, maybe it works :sweat_smile:

Reply:
Actually, this no longer works for some reason... I'm sure it worked at some point :/

Reply:
Ok got it, was failing with numpy bools. And your method works.

Reply:
So please add a comment with explanation, and best will be to add test covering.

Reply:
tests for the sorting you mean? Or for the general qt behaviour?

Can you give me a small example on how you would implement qt tests of this kind? (like, how would I test "sorting" works?)

Reply:
When I wrote this comment, The code part was marked as not covered by tests. You point that special code was needed for `np.bool`. It mean that for anyone in future that will touch this code, will not be warned by failing test if he wrongly refactored code. 

The `QTableView` has method `sortByColumn` https://doc.qt.io/qt-6/qtableview.html#sortByColumn.  As I understand, it should reorder items in model.

Reply:
Yep, should be good now with the test I just added I think!